platform/default/src/mbgl/util/thread_local.cpp

```diff
@@ -11,56 +8,37 @@
 
 namespace mbgl {
 namespace util {
-
-template <class T>
-class ThreadLocal<T>::Impl {
-public:
-    pthread_key_t key;
-};
-
-template <class T>
-ThreadLocal<T>::ThreadLocal() : impl(std::make_unique<Impl>()) {
-    int ret = pthread_key_create(&impl->key, [](void *) {});
-
-    if (ret) {
-        throw std::runtime_error("Failed to init local storage key.");
-    }
+namespace impl {
+
+ThreadLocalBase::ThreadLocalBase() {
+    static_assert(sizeof(storage) >= sizeof(pthread_key_t), "storage is too small");
+    static_assert(alignof(decltype(storage)) % alignof(pthread_key_t) == 0, "storage is incorrectly aligned");
+    if (const int ret = pthread_key_create(&reinterpret_cast<pthread_key_t&>(storage), nullptr)) {
+        throw std::runtime_error("Failed to init local storage key");
+    }    
 }
 
-template <class T>
-ThreadLocal<T>::~ThreadLocal() {
-    // ThreadLocal will not take ownership
-    // of the pointer it is managing. The pointer
-    // needs to be explicitly cleared before we
-    // destroy this object.
+ThreadLocalBase::~ThreadLocalBase() {
+    // ThreadLocal will not take ownership of the pointer it is managing. The pointer
+    // needs to be explicitly cleared before we destroy this object.
     assert(!get());
 
-    if (pthread_key_delete(impl->key)) {
-        Log::Error(Event::General, "Failed to delete local storage key.");
+    if (pthread_key_delete(reinterpret_cast<pthread_key_t&>(storage))) {
+        Log::Error(Event::General, "Failed to delete local storage key");
         assert(false);
     }
 }
 
-template <class T>
-T* ThreadLocal<T>::get() {
-    auto* ret = reinterpret_cast<T*>(pthread_getspecific(impl->key));
-    if (!ret) {
-        return nullptr;
-    }
-
-    return ret;
+void* ThreadLocalBase::get() {
+    return pthread_getspecific(reinterpret_cast<pthread_key_t&>(storage));
```

Comment:
pthread_getspecific may return null if it fails, should it be handled?

Reply:
It doesn't fail; `NULL` is a valid return value that indicates that the thread-specific variable hasn't been set. This is intended.

Reply:
:+1:  true, only error case is when invalid key is provided.